README.md
IP Subnet Calculator
=================

JavaScript class for calculating optimal subnet masks for non-standard IP ranges, e.g. 5.4.3.21 - 6.7.8.9

[View demo](http://franksrevenge.github.io/IPSubnetCalculator/)

[![Build Status](https://travis-ci.org/franksrevenge/IPSubnetCalculator.svg?branch=master)](https://travis-ci.org/franksrevenge/IPSubnetCalculator) [![Coverage Status](https://coveralls.io/repos/github/franksrevenge/IPSubnetCalculator/badge.svg?branch=master)](https://coveralls.io/github/franksrevenge/IPSubnetCalculator?branch=master) [![Dependency Status](https://david-dm.org/franksrevenge/IPSubnetCalculator/status.svg)](https://david-dm.org/franksrevenge/IPSubnetCalculator#info=dependencies&view=table) [![Dev Dependency Status](https://david-dm.org/franksrevenge/IPSubnetCalculator/dev-status.svg)](https://david-dm.org/franksrevenge/IPSubnetCalculator#info=devDependencies&view=table)


What does it do?
----------------

* Calculates subnet masks for standard and non-standard IP ranges. For example, `10.0.0.5 - 10.0.0.23` will result in `10.0.0.5/32, 10.0.0.6/31, 10.0.0.8/29, 10.0.0.16/29`.

* Calculates CIDR prefixes from subnet masks, e.g. `10.0.0.5/255.255.128.0` will result in `10.0.0.0/17`.

* Calculates subnet masks from CIDR prefixes, e.g. `10.0.0.5/17` will result in `255.255.128.0`.


Support
-------

* Node.js
* RequireJS
* Direct browser use


Installation
------------

```sh
> bower install ip-subnet-calculator

> npm install ip-subnet-calculator
```


Node.js
-------

```javascript
var IpSubnetCalculator = require( 'ip-subnet-calculator' );

console.log( IpSubnetCalculator.isIp( '127.0.0.1' ) ); // "true"
console.log( IpSubnetCalculator.toDecimal( '127.0.0.1' ) ); // "2130706433"

console.log( IpSubnetCalculator.calculate( '5.4.3.21', '6.7.8.9' ) );

```


RequireJS
---------

```javascript
require( [ 'ip-subnet-calculator' ],

function( IpSubnetCalculator )
{
    console.log( IpSubnetCalculator.isIp( '127.0.0.1' ) ); // "true"
    console.log( IpSubnetCalculator.toDecimal( '127.0.0.1' ) ); // "2130706433"
    
    console.log( IpSubnetCalculator.calculate( '5.4.3.21', '6.7.8.9' ) ); 
} );
```


Direct browser use
------------------

```html
<script src='lib/ip-subnet-calculator.js'></script>

<script>
    console.log( IpSubnetCalculator.isIp( '127.0.0.1' ) ); // "true"
    console.log( IpSubnetCalculator.toDecimal( '127.0.0.1' ) ); // "2130706433"
    
    console.log( IpSubnetCalculator.calculate( '5.4.3.21', '6.7.8.9' ) );
</script>
```


API
---

### IpSubnetCalculator.calculate( ipStart, ipEnd ) ###

Calculates an optimal set of IP masks for the given IP address range.

*ipStart* (string|number) Lowest IP in the range to be calculated in string (`123.123.123.0`) or numeric (`2071689984`) format.

*ipEnd* (string|number) Highest IP (inclusive) in the range to be calculated in string (`123.123.123.255`) or numeric (`2071690239`) format.

The function returns `null` in case of an error. Otherwise, an array containing one or more subnet
masks is returned:

```javascript
var result = [
    {
        ipLow              : 2071689984,
        ipLowStr           : "123.123.123.0",
        ipHigh             : 2071690239,
        ipHighStr          : "123.123.123.255",
        prefixMask         : 4294967040,
        prefixMaskStr      : "255.255.255.0",
        prefixSize         : 24,
        invertedMask       : 255,
        invertedMaskStr    : "0.0.0.255",
        invertedMaskSize   : 8
    },
    
    ...
];
```

Each object in question contain the following properties:

| Property             | Use                                                            |
:----------------------|:---------------------------------------------------------------|
| ipLow                | Decimal representation of the lowest IP address in the range   |
| ipLowStr             | String representation of the lowest IP address in the range    |
| ipHigh               | Decimal representation of the highest IP address in the range  |
| ipHighStr            | String representation of the highest IP address in the range   |
| prefixMask           | Decimal representation of the prefix (subnet) mask             |
| prefixMaskStr        | String representation of the prefix (subnet) mask              |
| prefixSize           | Size of the prefix (subnet) mask in bits                       |
| invertedMask         | Decimal representation of the inverted prefix mask             |
| invertedMaskStr      | String representation of the inverted prefix mask              |
| invertedSize         | Size of the inverted prefix max in bits                        |


### IpSubnetCalculator.calculateSubnetMask( ip, prefixSize ) ###

Calculates a subnet mask from CIDR prefix.

*ip* (string|number) IP address in string or numeric format

*prefixSize* Number of relevant bits in the subnet mask

The function returns an object containing full description of the IP range, as described in `IpSubnetCalculator.calculate()`.


### IpSubnetCalculator.calculateCIDRPrefix( ip, subnetMask ) ###

Calculates a CIDR prefix from subnet mask.

*ip* (string|number) IP address in string or numeric format

*subnetMask* (string|number) IP subnet mask in string or numeric format

The function returns an object containing full description of the IP range, as described in `IpSubnetCalculator.calculate()`.


## Test Functions ##

### IpSubnetCalculator.isIp( ipStr ) ###

Tests whether string is an IP address.

*ipStr* (string) A string

The function returns a `true` if the string is an IP address, `false` otherwise.

### IpSubnetCalculator.isDecimalIp( ipNum ) ###

Tests whether `ipNum` is a decimal IP address.

*ipNum* (number) A number

The function returns a `true` if the number is an IP address, `false` otherwise.


## Conversion Functions ##

### IpSubnetCalculator.toDecimal( ip ) ###

Calculates a decimal integer from an string IP address.

*ip* (string|number) IP address in string format

The function returns a decimal representation of an IP address as an integer. If a valid numeric representation 
of an IP is passed to this function, it is returned unmodified.

If an invalid value is passed to the function, it will `throw` an `Error` object.


### IpSubnetCalculator.toString( num ) ###

*num* (number|string) Decimal representation of an IP address.

The function returns an IP address as a string. If a valid string representation of an IP is passed to this function,
it is returned unmodified.

If an invalid value is passed to the function, it will `throw` an `Error` object.


License
-------

[MIT](http://opensource.org/licenses/MIT)




CHANGES.md
1.1.7
-----

* Added LICENSE file


1.1.0
------

* `isIp()` and `isDecimalIp()` functions
* Added tests
* Added `.editorconfig`, `.eslintrc`, and `.jshintrc` files
* `calculate()`, `calculateSubnetMask()`, and `calculateCIDRPrefix()` now accept IPs in string and number format


1.0.0
-----

Initial release


test/ip-subnet-calculator.js
'use strict';

var should	= require( 'should' ),
	ipsc	= require( '../lib/ip-subnet-calculator' ),
	_		= require( 'lodash' );


var ipList = {
	'127.0.0.1'			: 2130706433,
	'8.8.4.4'			: 134743044,
	'8.8.8.8'			: 134744072,
	'4.2.2.2'			: 67240450,
	'255.255.255.255'	: 4294967295,
	'0.0.0.0'			: 0
};


var ipValidity = {
	'127.0.0.1'			: true,
	'255.255.255.255'	: true,
	'0.0.0.0'			: true,
	'12.32'				: false,
	'0.0.0.0.1'			: false,
	'123.123.512.123'	: false,
	' 123.123.123.123 '	: false,
	'hello world'		: false,
	'#0'				: false,
	'#67240450'			: false
};


var ipNumValidity = {

	'#2130706433'	: true,
	'#134743044'	: true,
	'#134744072'	: true,
	'#67240450'		: true,
	'#4294967295'	: true,
	'#0'			: true,
	'#4294967296'	: false,
	'#11234.5839'	: false,
	'hello world'	: false,
	'127.0.0.1'		: false
};


var ipRanges = [
	{
		start	: '10.0.0.0',
		end		: '10.0.0.255',
		mask	: 24,
		success	: true
	},
	{
		start	: '128.0.0.0',
		end		: '128.0.255.255',
		mask	: 16,
		success	: true
	},
	{
		start	: '10.0.0.0',
		end		: '10.0.0.0',
		mask	: 32,
		success	: true
	},
	{
		start	: '0.0.0.0',
		end		: '255.255.255.255',
		mask	: 0,
		success	: true
	},
	{
		start	: '10.5.0.0',
		end		: '10.0.3.25',
		success	: false
	}
];


var ipSubnets = [

	{
		ip		: '10.0.0.45',
		mask	: 24,
		start	: '10.0.0.0',
		end		: '10.0.0.255'
	},
	{
		ip		: '101.0.2.45',
		mask	: 16,
		start	: '101.0.0.0',
		end		: '101.0.255.255'
	}

];





describe( 'IP Subnet Calculator',
	function()
	{
		it( 'should convert string IPs to decimals',
			function()
			{
				_.each(
						ipList,
						function( ipDec, ipStr )
						{
							ipsc.toDecimal( ipStr ).should.be.exactly( ipDec );
						}
					);
			}
		);


		it( 'should convert decimal IPs to strings',
				function()
				{
					_.each(
							ipList,
							function( ipDec, ipStr )
							{
								ipsc.toString( ipDec ).should.be.exactly( ipStr );
							}
						);
				}
		);
		
		
		it( 'should correctly create bitmasks',
				function()
				{
					ipsc.getMask( 0 ).should.be.exactly( 0 );
					ipsc.getMask( 4 ).should.be.exactly( 15 );
					ipsc.getMask( 15 ).should.be.exactly( 32767 );
					ipsc.getMask( 32 ).should.be.exactly( 4294967295 );
				}
			);


		it( 'should correctly create prefix bitmasks',
				function()
				{
					ipsc.getPrefixMask( 0 ).should.be.exactly( 0 );
					ipsc.getPrefixMask( 4 ).should.be.exactly( 4026531840 );
					ipsc.getPrefixMask( 15 ).should.be.exactly( 4294836224 );
					ipsc.getPrefixMask( 32 ).should.be.exactly( 4294967295 );
				}
		);


		it( 'should validate IP addresses',
				function()
				{
					_.each(
							ipValidity,
							function( isValid, ipStr )
							{
								if( ipStr.substr( 0, 1 ) === '#' )
								{
									ipStr = parseFloat( ipStr.substr( 1 ) );
								}

								ipsc.isIp( ipStr ).should.be.exactly( isValid );
							}
					);
				}
			);


		it( 'should validate numeric IP addresses',
				function()
				{
					_.each(
							ipNumValidity,
							function( isValid, ipNum )
							{
								if( ipNum.substr( 0, 1 ) === '#' )
								{
									ipNum = parseFloat( ipNum.substr( 1 ) );
								}

								ipsc.isDecimalIp( ipNum ).should.be.exactly( isValid );
							}
					);
				}
		);


		it( 'should calculate optimal range for a maskable range',
				function()
				{
					_.each(
							ipRanges,
							function( ipRange )
							{
								var range = ipsc.getOptimalRange( ipsc.toDecimal( ipRange.start ), ipsc.toDecimal( ipRange.end ) );

								if( ipRange.success )
								{
									range.ipLowStr.should.be.exactly( ipRange.start );
									range.ipLow.should.be.exactly( ipsc.toDecimal( range.ipLowStr ) );
									range.ipHighStr.should.be.exactly( ipRange.end );
									range.ipHigh.should.be.exactly( ipsc.toDecimal( range.ipHighStr ) );
									range.prefixSize.should.be.exactly( ipRange.mask );
									range.invertedSize.should.be.exactly( 32 - ipRange.mask );
									range.invertedMask.should.be.exactly( ipsc.getMask( range.invertedSize ) );
									range.prefixMask.should.be.exactly( ipsc.getPrefixMask( range.prefixSize ) );
									range.invertedMaskStr.should.be.exactly( ipsc.toString( range.invertedMask ) );
									range.prefixMaskStr.should.be.exactly( ipsc.toString( range.prefixMask ) );
								}
								else
								{
									should.equal( range, null );
								}
							}
						);
				}
			);


		it( 'should calculate CIDR masks',
				function()
				{
					_.each(
							ipSubnets,
							function( subnet )
							{
								var range = ipsc.calculateCIDRPrefix(
										subnet.ip,
										ipsc.toString( ipsc.getPrefixMask( subnet.mask ) )
									);

								range.ipLowStr.should.be.exactly( subnet.start );
								range.ipHighStr.should.be.exactly( subnet.end );
								range.prefixSize.should.be.exactly( subnet.mask );
								range.invertedSize.should.be.exactly( 32 - subnet.mask );
							}
					);
				}
		);


		it( 'should calculate subnet masks',
				function()
				{
					_.each(
							ipSubnets,
							function( subnet )
							{
								var range = ipsc.calculateSubnetMask( subnet.ip, subnet.mask );

								range.ipLowStr.should.be.exactly( subnet.start );
								range.ipHighStr.should.be.exactly( subnet.end );
								range.prefixSize.should.be.exactly( subnet.mask );
								range.invertedSize.should.be.exactly( 32 - subnet.mask );
							}
						);
				}
			);


		it( 'should calculate masks for non-standard ranges',
				function()
				{
					var ranges = ipsc.calculate( '10.0.0.0', '10.0.1.1' );

					ranges[ 0 ].ipLowStr.should.be.exactly( '10.0.0.0' );
					ranges[ 0 ].ipHighStr.should.be.exactly( '10.0.0.255' );
					ranges[ 1 ].ipLowStr.should.be.exactly( '10.0.1.0' );
					ranges[ 1 ].ipHighStr.should.be.exactly( '10.0.1.1' );
				}
			);


		it( 'should calculate masks for non-standard ranges',
				function()
				{
					var ranges = ipsc.calculate( '10.0.0.0', '10.0.1.1' );

					ranges[ 0 ].ipLowStr.should.be.exactly( '10.0.0.0' );
					ranges[ 0 ].ipHighStr.should.be.exactly( '10.0.0.255' );
					ranges[ 1 ].ipLowStr.should.be.exactly( '10.0.1.0' );
					ranges[ 1 ].ipHighStr.should.be.exactly( '10.0.1.1' );
				}
		);


		it( 'should reject requests with invalid IPs',
				function()
				{
					should.equal( ipsc.calculate( 'hello', 'world' ), null );
				}
			);


		it( 'should fail calculate() if start IP is higher than end IP',
				function()
				{
					should.equal( ipsc.calculate( '10.1.0.0', '10.0.0.0' ), null );
				}
		);


		it( 'should fail calculateSubnetMask() if IP is invalid',
				function()
				{
					should.equal( ipsc.calculateSubnetMask( 'helloworld', 24 ), null );
				}
		);


		it( 'should fail calculateCIDRPrefix() if IP is invalid',
				function()
				{
					should.equal( ipsc.calculateCIDRPrefix( 'helloworld', '255.255.255.255' ), null );
				}
		);


		it( 'should pass decimal IPs through toDecimal()',
				function()
				{
					ipsc.toDecimal( 4026531840 ).should.be.exactly( 4026531840 );
				}
			);


		it( 'should pass string IPs through toString()',
				function()
				{
					ipsc.toString( '127.0.0.1' ).should.be.exactly( '127.0.0.1' );
				}
		);


		it( 'should reject converting invalid numeric IPs to strings',
				function()
				{
					(
						function()
						{
							ipsc.toString( -1 );
						}
					)
					.should.throw();
				}
		);

	}
);




resources/js/ui.js

require( [ 'jquery', 'ip-subnet-calculator' ],

function( $, IpSubnetCalculator )
{
	'use strict';

	var showResult = function( elementName, msg )
	{
		$( '#' + elementName ).html( msg );
		$( '#' + elementName ).slideDown( 'fast' );
	};


	$( '#ip_calculate_range_button' ).click(
			function()
			{
				var result = IpSubnetCalculator.calculate(
						$( '#ip_calculate_range_start_ip' ).val(),
						$( '#ip_calculate_range_end_ip' ).val()
					);

				if( result === null )
				{
					showResult( 'ip_calculate_range_result', 'Please enter IP range in the fields above' );
				}
				else
				{
					var i;
					var msg = '<h2>Result</h2>' +
								'<table>' +
								'<thead><tr><td>IP</td><td>Prefix</td><td>Subnet Mask</td></tr></thead>' +
								'<tfoot><tr><td>IP</td><td>Prefix</td><td>Subnet Mask</td></tr></tfoot>';

					for( i = 0; i < result.length; i++ )
					{
						msg += '<tr>' +
									'<td>' + result[ i ].ipLowStr + '</td>' +
									'<td>' + '/' + result[ i ].prefixSize + '</td>' +
									'<td>' + result[ i ].prefixMaskStr + '</td>' +
									"</tr>\n";
					}

					msg += "</tbody></table>\n";

					showResult( 'ip_calculate_range_result', msg );
				}
			}
		);


	$( '#ip_calculate_cidr_mask_button' ).click(
			function()
			{
				var result = IpSubnetCalculator.calculateSubnetMask(
						$( '#ip_calculate_cidr_mask_ip' ).val(),
						$( '#ip_calculate_cidr_mask_mask' ).val()
					);

				if( result === null )
				{
					showResult( 'ip_calculate_cidr_mask_result', 'Please enter IP range in the fields above' );
				}
				else
				{
					var msg = '<h2>Result</h2>' +
								'<table>' +
								'<thead><tr><td>Address Range</td><td>Subnet Mask</td><td>Prefix Size</td></tr></thead>' +
								'<tbody>' +
								'<tr>' +
									'<td>' + result.ipLowStr + ' - ' + result.ipHighStr + '</td>' +
									'<td>' + result.prefixMaskStr + '</td>' +
									'<td>' + result.prefixSize + '</td>' +
								'</tr>' +
								'</tbody></table>';

					showResult( 'ip_calculate_cidr_mask_result', msg );
				}
			}
		);


	$( '#ip_calculate_mask_button' ).click(
			function()
			{
				var result = IpSubnetCalculator.calculateCIDRPrefix(
						$( '#ip_calculate_mask_ip' ).val(),
						$( '#ip_calculate_mask_mask' ).val()
					);

				if( result === null )
				{
					showResult( 'ip_calculate_mask_result', 'Please enter IP range in the fields above' );
				}
				else
				{
					var msg = '<h2>Result</h2>' +
								'<table>' +
								'<thead><tr><td>Address Range</td><td>Subnet Mask</td><td>Prefix Size</td></tr></thead>' +
								'<tbody>' +
								'<tr>' +
									'<td>' + result.ipLowStr + ' - ' + result.ipHighStr + '</td>' +
									'<td>' + result.prefixMaskStr + '</td>' +
									'<td>' + result.prefixSize + '</td>' +
								'</tr>' +
								'</tbody></table>';

					showResult( 'ip_calculate_mask_result', msg );
				}
			}
		);

} );



IPSubnetCalculator.js
/*jshint strict: false, bitwise: false */


/**
* \class IPSubnetCalculator
* 
* IPSubnetCalculator 0.1.0
* 
* Copyright (c) 2013, Aleksi Asikainen   
* All rights reserved.
* 
* Released under Apache 2.0 License
* http://www.apache.org/licenses/LICENSE-2.0.html 
* 
*  
* A singleton class for:
* 
*    1) Calculating optimal and exact subnet masks for an
*       unrestricted range of IP addresses.
* 
*       E.g. range 10.0.1.255 - 10.0.3.255 should result in:  
*  
*           10.0.1.255/32
*           10.0.2.0/24
*           10.0.3.0/24 
*   
*    2) Calculating subnets from an IP and bitmask size
*    
*    3) Calculating subnets from an IP and subnet mask   
*
* 
* Use calculate(), calculateCIDRMask(), and calculateMask() for easy access.
*
**/


var IPSubnetCalculator = {

	/**
	* \fn calculate( ipStart, ipEnd )
	*
	* Calculates an optimal set of IP masks for the given IP address range	 
	*
	* \param ipStart String Lowest IP in the range to be calculated in string format ("123.123.123.123")  
	* \param ipEnd String Highest IP (inclusive) in the range to be calculated in string format ("123.123.123.123")
	*
	* \return The function returns null in case of an error. Otherwise, an array containing one or more subnet 
	*         masks is returned:
	*
	*         [
	*             {
	*                 ipLow              : 2071689984,
	*                 ipLowStr           : "123.123.123.0",
	*                 ipHigh             : 2071690239,
	*                 ipHighStr          : "123.123.123.255",
	*                 prefixMask         : 4294967040,
	*                 prefixMaskStr      : "255.255.255.0",
	*                 prefixSize         : 24,
	*                 invertedMask       : 255,
	*                 invertedMaskStr    : "0.0.0.255",
	*                 invertedMaskSize   : 8
	*             },
	*
	*             ...
	*
	*         ]
	*
	**/

	calculate : function IPSubnetCalculator_calculate( ipStart, ipEnd )
	{
		var ipStartNum, ipEndNum, ipCurNum;
		var rangeCollection = [];
				
		if( 
			( ipStart === '' ) || ( ipStart === null ) || ( ipStart === false ) || 
			( ipEnd === '' ) || ( ipEnd === null ) || ( ipEnd === false ) 
		)
		{
			return null;
		}
		
		
		ipStartNum = IPSubnetCalculator.toDecimal( ipStart );
		ipEndNum = IPSubnetCalculator.toDecimal( ipEnd );
		
		if( ipEndNum < ipStartNum )
		{
			return null;
		}
	
		ipCurNum = ipStartNum;
		
		while( ipCurNum <= ipEndNum )
		{
			var optimalRange = IPSubnetCalculator.getOptimalRange( ipCurNum, ipEndNum );
			
			if( optimalRange === null )
			{
				return null;
			}
		
			rangeCollection.push( optimalRange );
			
			ipCurNum = optimalRange.ipHigh + 1;		
		}
				
		return rangeCollection;	
	},
	
	
	/**
	* \fn calculateCIDRMask( ip, prefixSize )
	* \param ip IP address in string format
	* \param prefixSize Number of relevant bits in the subnet mask
	* \return Returns null in case of an error, and a subnet data object otherwise.
	*         For details about the subnet data object, see documentation of
	*         getMaskRange()	 
	**/
	
	calculateCIDRMask : function IPSubnetCalculator_calculateCIDRMask( ip, prefixSize )
	{
		if( 
			( ip === '' ) || ( ip === null ) || ( ip === false ) || 
			( prefixSize === '' ) || ( prefixSize === null ) || ( prefixSize === false ) 
		)
		{
			return null;
		}

	
		var ipNum = IPSubnetCalculator.toDecimal( ip );
				
		return IPSubnetCalculator.getMaskRange( ipNum, prefixSize );	
	},
	
	
	/**
	* \fn calculateMask( ip, ipMask )
	* \param ip IP address in string format
	* \param ipMask IP subnet mask in string format ("255.255.255.0")
	* \return Returns null in case of an error, and a subnet data object otherwise.
	*         For details about the subnet data object, see documentation of
	*         getMaskRange()	 
	**/
	
	calculateMask : function IPSubnetCalculator_calculateMask( ip, ipMask )
	{
		if( 
			( ip === '' ) || ( ip === null ) || ( ip === false ) || 
			( ipMask === '' ) || ( ipMask === null ) || ( ipMask === false ) 
		)
		{
			return null;
		}

	
		var ipNum = IPSubnetCalculator.toDecimal( ip );
		var ipMaskNum = IPSubnetCalculator.toDecimal( ipMask );
	
		var prefix = 0;
		var newPrefix = 0;
		var prefixSize;
	
		for( prefixSize = 0; prefixSize < 32; prefixSize++ )
		{
			newPrefix = ( prefix + ( 1 << ( 32 - ( prefixSize + 1 ) ) ) ) >>> 0;
		
			if( ( ( ipMaskNum & newPrefix ) >>> 0 ) !== newPrefix )
			{
				break;
			}
		
			prefix = newPrefix;		
		} 
		
		return IPSubnetCalculator.getMaskRange( ipNum, prefixSize );		 
	},
	

	/**
	* \fn getOptimalRange( ipNum, ipEndNum )
	* 
	* Finds the largest subnet mask that begins from ipNum and does not
	* exceed ipEndNum.
	* 
	* \return Returns null on failure, otherwise an object with the following fields:
	* 
	* ipLow - Decimal representation of the lowest IP address in the subnet
	* ipLowStr - String representation of the lowest IP address in the subnet
	* ipHigh - Decimal representation of the highest IP address in the subnet
	* ipHighStr - String representation of the highest IP address in the subnet
	* prefixMask - Bitmask matching prefixSize
	* prefixMaskStr - String / IP representation of the bitmask
	* prefixSize - Size of the prefix
	* invertedMask - Bitmask matching the inverted subnet mask
	* invertedMaskStr - String / IP representation of the inverted mask
	* invertedSize - Number of relevant bits in the inverted mask
	*
	*/

	getOptimalRange : function IPSubnetCalculator_getOptimalRange( ipNum, ipEndNum )
	{
		var prefixSize;
		var optimalRange = null;
		
		for( prefixSize = 32; prefixSize >= 0; prefixSize-- )
		{
			var maskRange = IPSubnetCalculator.getMaskRange( ipNum, prefixSize );
			
			if( ( maskRange.ipLow === ipNum ) && ( maskRange.ipHigh <= ipEndNum ) )
			{
				optimalRange = maskRange;
			}
			else
			{
				break;
			}
		}
		
		return optimalRange;
	}, 


	/**
	* \fn getMaskRange( ipNum, prefixSize )
	* \param ipNum Decimal IP address
	* \param prefixSize Subnet mask size in bits
	* 
	* Calculates details of a CIDR subnet
	* 
	* \return Returns an object with the following fields:
	* 
	* ipLow - Decimal representation of the lowest IP address in the subnet
	* ipLowStr - String representation of the lowest IP address in the subnet
	* ipHigh - Decimal representation of the highest IP address in the subnet
	* ipHighStr - String representation of the highest IP address in the subnet
	* prefixMask - Bitmask matching prefixSize
	* prefixMaskStr - String / IP representation of the bitmask
	* prefixSize - Size of the prefix
	* invertedMask - Bitmask matching the inverted subnet mask
	* invertedMaskStr - String / IP representation of the inverted mask
	* invertedSize - Number of relevant bits in the inverted mask
	*
	**/

	getMaskRange : function IPSubnetCalculator_getMaskRange( ipNum, prefixSize )
	{
		var prefixMask = IPSubnetCalculator.getPrefixMask( prefixSize );
		var lowMask = IPSubnetCalculator.getMask( 32 - prefixSize );
		
		var ipLow = ( ipNum & prefixMask ) >>> 0;
		var ipHigh = ( ( ( ipNum & prefixMask ) >>> 0 ) + lowMask ) >>> 0;
		
		return {
			'ipLow'				: ipLow,
			'ipLowStr'			: IPSubnetCalculator.toString( ipLow ),
			
			'ipHigh'			: ipHigh,
			'ipHighStr'			: IPSubnetCalculator.toString( ipHigh ),
			
			'prefixMask'		: prefixMask,
			'prefixMaskStr'		: IPSubnetCalculator.toString( prefixMask ),
			'prefixSize'		: prefixSize,
			
			'invertedMask'		: lowMask,
			'invertedMaskStr'	: IPSubnetCalculator.toString( lowMask ),
			'invertedSize'		: 32 - prefixSize
		};
	},


	/**
	* \fn getPrefixMask( prefixSize )
	* \param prefixSize Number of bits to be set	 
	* 
	* Creates a bitmask with maskSize leftmost bits set to one 
	* 
	* \return Returns the bitmask
	*
	**/
	
	getPrefixMask : function( prefixSize )
	{
		var mask = 0;
		var i;
		
		for( i = 0; i < prefixSize; i++ )
		{
			mask += ( 1 << ( 32 - ( i + 1 ) ) ) >>> 0;
		}
		
		return mask;
	},
	
	
	/**
	* \fn getMask( maskSize )
	* \param maskSize Number of bits to be set	 
	* 
	* Creates a bitmask with maskSize rightmost bits set to one
	* 
	* \return Returns the bitmask
	*
	**/
	
	getMask : function( maskSize )
	{
		var mask = 0;
		var i;
		
		for( i = 0; i < maskSize; i++ )
		{
			mask += ( 1 << i ) >>> 0;
		}
		
		return mask;
	},
	
	
	/**
	* \fn toDecimal( ipString )
	* \param ipString IP address in string format	 
	* 
	* Converts string formatted IPs to decimal representation
	*
	* Function copied from http://javascript.about.com/library/blipconvert.htm	
	*
	* \return Returns the IP address in decimal format
	*
	**/
	
	toDecimal : function IPSubnetCalculator_toDecimal( ipString )
	{
		var d = ipString.split( '.' );
		return ( ( ( ( ( ( + d[ 0 ] ) * 256 ) + ( +d [ 1 ] ) ) * 256 ) + ( +d[ 2 ] ) ) * 256 ) + ( +d[ 3 ] );
	},
	
	
	/**
	* \fn toString( ipNum )
	* \param ipNum IP address in decimal format
	* 
	* Converts decimal IPs to string representation
	*
	* Function copied from http://javascript.about.com/library/blipconvert.htm
	* 
	* \return Returns the IP address in string format
	*
	**/
	
	toString : function IPSubnetCalculator_toString( ipNum )
	{	
		var d = ipNum % 256;
		
		for( var i = 3; i > 0; i-- )
		{ 
			ipNum = Math.floor( ipNum / 256 );
			d = ipNum % 256 + '.' + d;
		}
		
		return d;
	}



};



